gazelle/gazelle.go

```diff
@@ -256,3 +257,52 @@ func getBzlFileLoads(path string) ([]string, error) {
 
 	return loads, nil
 }
+
+func isBzlSourceFile(f string) bool {
+	return strings.HasSuffix(f, fileType) && !ignoreSuffix.Matches(f)
+}
+
+// generateEmpty generates the list of rules that don't need to exist in the
+// BUILD file any more.
+// For each bzl_library rule in args.File that only has srcs that aren't in
+// args.RegularFiles or args.GenFiles, add a bzl_library with no srcs or deps.
+// That will let Gazelle delete bzl_library rules after the corresponding .bzl
+// files are deleted.
+func generateEmpty(args language.GenerateArgs) []*rule.Rule {
+	var ret []*rule.Rule
+	if args.File == nil {
+		return ret
+	}
+	for _, r := range args.File.Rules {
+		if r.Kind() != "bzl_library" {
+			continue
+		}
+		name := r.AttrString("name")
+		srcs := srcsList(r.AttrStrings("srcs"))
+
+		for _, f := range args.RegularFiles {
```

Comment:
Instead of defining `srcsList` below, I think it would be better to keep the logic inline.

Time complexity could be lowered by building a set of existing files first. Something like: 

```
exists := make(map[string]bool)
for _, f := range args.RegularFiles {
  exists[f] = true
}
for _, f := range args.GenFiles {
  exists[f] = true
}
for _, r := range args.File.Rules {
  srcsExist := false
  for _, f := range f.AttrStrings("srcs") {
    if exists[f] {
      srcsExist = true
      break
    }
  }
  if !srcsExist {
    ret = append(ret, rule.NewRule("bzl_library", name))
  }
}
```